### PR TITLE
[Feature] Add prometheus and component shutdown logging

### PR DESCRIPTION
**Pull request by zquestz**
This is the patch set I have been running on my server for a while now.

- Adds prometheus and cleans up a few bugs in the previous PR.
- Adds server logging for all component shutdowns so we can debug shutdown issues easier.

### CODE REVIEW

File: grpcserver.go [hunk 1]
```diff
@@ -3,22 +3,29 @@ package main
...
-	"net/http"
-	"strings"
 )
 
 // AuthenticationTokenKey is the key used in the context to authenticate clients.
 // If this is set to anything other than "" in the config, then the server expects
 // the client to set a key value in the context metadata to 'AuthenticationToken: cfg.AuthToken'
 const AuthenticationTokenKey = "AuthenticationToken"
 
+var prometheusEnabled = false
```
Comment (Ekliptor) [inline grpcserver.go:27]: the logging is nice. But this line is ugly. Variables shouldn't be set globally in a file. Instead bound to a struct (such as `rpcCfg *bchrpc.GrpcServerConfig`)

**Review by Ekliptor — COMMENTED** [review]
the logging is nice. But this line is ugly. Variables shouldn't be set globally in a file. Instead bound to a struct (such as `rpcCfg *bchrpc.GrpcServerConfig`)

File: grpcserver.go (hunk 1 above)
Comment (zquestz) [reply grpcserver.go:27]: So your original PR broke all RPC calls if prometheus wasn't specified, plus those interceptors don't have access to the config. If you have a better pattern to check if prometheus is enabled, I am happy to change it.

Comment (Ekliptor) [reply grpcserver.go:27]: Doing a cast from `interface{}`

Comment (zquestz) [reply grpcserver.go:27]: So you prefer doing a type check on every grpc request vs a package level var that just tracks if prometheus has been enabled? I agree that globals aren't great, but doing type checks on every grpc request seems expensive...

Comment (Ekliptor) [reply grpcserver.go:27]: Hmm agreed - didn't benchmark and generally type assertions are fast in Go (unlike type switch). But still even without benchmark we can agree your solution is faster and less code. We won't start a 2nd gRPC server ever where the global gets in the way I guess

**Pull request merged**